src/hooks/sounds.tsx

```diff
@@ -0,0 +1,164 @@
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from "react";
+
+import useSound from "use-sound";
+import {
+  ExposedData,
+  HookOptions,
+  PlayFunction,
+  PlayOptions,
+} from "use-sound/dist/types";
+import { HowlOptions } from "howler";
+
+import { SoundConfigMap, SoundConfigReference } from "types/sounds";
+
+import { fetchSoundConfigs } from "api/sounds";
+
+export type PlaySpriteFunction = (options?: PlaySpriteOptions) => void;
+export type PlaySpriteOptions = Omit<PlayOptions, "id">;
+
+export type UseCustomSoundOptions = HookOptions & HowlOptions;
+
+export interface ExposedDataWithPlay extends ExposedData {
+  play: PlayFunction;
+}
+
+export interface CustomSoundsState {
+  soundConfigs: SoundConfigMap;
+}
+
+export const initialValue: CustomSoundsState = {
+  soundConfigs: {},
+};
+
+export const CustomSoundsContext = createContext<CustomSoundsState>(
+  initialValue
+);
+
+export const CustomSoundsProvider: React.FC = ({ children }) => {
+  const [soundConfigs, setSoundConfigs] = useState<SoundConfigMap>(
+    initialValue.soundConfigs
+  );
+
+  // Fetch the sound configs data on first load
+  useEffect(() => {
+    fetchSoundConfigs().then((soundConfigs) => {
+      setSoundConfigs(soundConfigs);
+    });
+  }, []);
+
+  const providerData = useMemo(
+    () => ({
+      soundConfigs,
+    }),
+    [soundConfigs]
+  );
+
+  return (
+    <CustomSoundsContext.Provider value={providerData}>
+      {children}
+    </CustomSoundsContext.Provider>
+  );
+};
+
+export const useCustomSoundsContext = (): CustomSoundsState =>
+  useContext(CustomSoundsContext);
+
+const USE_SOUND_DISABLED_URL = "";
+const USE_SOUND_DISABLED_CONFIG = { soundEnabled: false };
+
+/**
+ * Load a custom sound using a SoundConfigReference, and expose controls to play it/a sprite within it.
+ *
+ * @example
+ *   const [play] = useCustomSound({ soundId: "ABC123" });
+ *
+ * @example
+ *   const [play] = useCustomSound({
+ *     soundId: "ABC123",
+ *     spriteName: "somethingSnazzy",
+ *   });
+ *
+ * @param soundRef
+ * @param options see use-sound's HookOptions
+ *
+ * @see https://github.com/joshwcomeau/use-sound
+ * @see HookOptions (from use-sound)
+ * @see ReturnedValue (from use-sound)
+ * @see PlayFunction (from use-sound)
+ * @see ExposedData (from use-sound)
+ *
+ * @see https://github.com/goldfire/howler.js
+ * @see HowlOptions
+ */
+export const useCustomSound = (
+  soundRef?: string | SoundConfigReference,
+  options?: UseCustomSoundOptions
+): [PlaySpriteFunction, ExposedDataWithPlay] => {
+  const { soundId, spriteName } = (() => {
+    switch (typeof soundRef) {
+      case "string":
+        return { soundId: soundRef, spriteName: undefined };
+
+      // This case won't actually allow us to play any sounds, and is only here to
+      // work around the inability to conditionally render hooks
+      case "undefined":
+        return { soundId: "", spriteName: undefined };
+
+      default:
+        return soundRef;
+    }
+  })();
```

Comment:
It took me a while to actually understand what is going on here. Is it possible to reduce the complexity. Maybe just use a ternary operator, if you don't plan on expanding the data types of `soundRef`.

Reply:
There are 3 distinct cases that need to be handled here, how would you see that being simplified as a ternary?

This is a pretty standard usage of a switch/case statement. The IIFE (immediately invoked function expression) is also fairly standard (though perhaps less commonly used these days) An alternative to that would be to pull it out into a separate function, but that felt overkill given its’s only ever useful/needed in this one place.

Reply:
What do you mean by “don’t plan on expanding the data types of sound ref”?

Reply:
[FYI] As an aside, the pattern I've used here is sort of a hack/workaround that allows us to do things in a similar way to the ECMAScript Pattern Matching proposal (though not as powerful as it), that is currently in 'Stage 1' of the tc39 system:

> Stage 1 proposals represent problems that the committee is interested in spending time exploring solutions to.

- https://github.com/tc39/proposal-pattern-matching

There is also a PR for a babel plugin (handles our transpilation) that implements this, but it seems to still be outstanding, pending expected changes to the design of the spec:

- https://github.com/babel/babel/pull/9318